File: frontend/src/api/indicadores.ts
import { api } from './client'

export type KpiKey = 'tasa_servicio' | 'disponibilidad' | 'eficiencia' | 'calidad'

export interface PeriodoIndicador {
  inicio: string
  fin: string
  mes_label: string
}

export interface SemanaValor {
  semana_label: string
  inicio: string
  fin: string
  valor: number
  estado?: 'pasada' | 'en_curso' | 'futura'
}

export interface MaquinaValor {
  maquina_id: number
  maquina_nombre: string | null
  valor: number
  // campos detallados opcionales por KPI
  dias_trabajados?: number | null
  horas_disponibles?: number | null
  horas_parada?: number | null
  horas_operativas?: number | null
  capacidad_hora?: number | null
  produccion_real?: number | null
  produccion_teorica?: number | null
  produccion_buena?: number | null
  clase_b?: number | null
  desecho?: number | null
  produccion_total?: number | null
  total_ops?: number | null
  ops_atrasadas?: number | null
}

export interface IndicadorData {
  kpi: KpiKey
  periodo: PeriodoIndicador
  meta: number | null
  valor_periodo: number
  por_semana: SemanaValor[]
  por_maquina: MaquinaValor[]
}

export const fetchIndicador = (
  kpi: KpiKey,
  mes?: string,
  maquinaId?: number,
): Promise<IndicadorData> => {
  const params: Record<string, string | number> = {}
  if (mes) params.mes = mes
  if (maquinaId != null) params.maquina_id = maquinaId
  return api.get(`/api/indicadores/${kpi}`, { params }).then(r => r.data)
}

File: backend/schemas/indicadores.py
from datetime import date
from typing import List, Literal, Optional
from pydantic import BaseModel

EstadoOp = Literal["A tiempo", "Completada tarde", "Atrasada", "En plazo", "Completada"]


class PeriodoIndicadorOut(BaseModel):
    inicio: date
    fin: date
    mes_label: str  # ej. "Mayo 2026"


class SemanaValorOut(BaseModel):
    semana_label: str  # ej. "Sem 21 (18-22 May)"
    inicio: date
    fin: date
    valor: float
    estado: Literal["pasada", "en_curso", "futura"] = "pasada"


class MaquinaValorOut(BaseModel):
    maquina_id: int
    maquina_nombre: Optional[str] = None
    valor: float
    # Campos detallados por KPI (opcionales — se rellenan según el KPI)
    dias_trabajados: Optional[int] = None
    # disponibilidad
    horas_disponibles: Optional[float] = None
    horas_parada: Optional[float] = None
    # eficiencia
    horas_operativas: Optional[float] = None
    capacidad_hora: Optional[int] = None
    produccion_real: Optional[int] = None
    produccion_teorica: Optional[float] = None
    # calidad
    produccion_buena: Optional[int] = None
    clase_b: Optional[int] = None
    desecho: Optional[int] = None
    produccion_total: Optional[int] = None
    # tasa_servicio
    total_ops: Optional[int] = None
    ops_atrasadas: Optional[int] = None
    # oee (componentes: valor lleva el OEE%)
    disponibilidad_pct: Optional[float] = None
    rendimiento_pct: Optional[float] = None
    calidad_pct: Optional[float] = None


class OpTasaServicioOut(BaseModel):
    op_docto: int
    item: Optional[str] = None
    referencia: Optional[str] = None      # ext1 (referencia/color)
    marca: Optional[str] = None
    maquina_id: Optional[int] = None
    maquina_nombre: Optional[str] = None
    fecha_prometida: date
    fecha_completada: Optional[date] = None  # fecha del último registro de producción
    cantidad: int
    cant_consumida: int
    pct_completado: float
    estado: EstadoOp
    dias_atraso: Optional[int] = None     # días de retraso (positivo = retrasado)


class IndicadorOut(BaseModel):
    kpi: str  # tasa_servicio | disponibilidad | eficiencia | calidad
    periodo: PeriodoIndicadorOut
    meta: Optional[float] = None
    valor_periodo: float
    por_semana: List[SemanaValorOut]
    por_maquina: List[MaquinaValorOut]
